<a href="https://colab.research.google.com/github/Sanyabansal115/public-safety-model-training/blob/Sanya/KSI_Group5_section_001COMP247Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [8]:
import pandas as pd
import numpy as np

In [10]:
ksi_df = pd.read_csv("TOTAL_KSI_4115794401574937330.csv")
ksi_df.head()

,OBJECTID,INDEX,ACCNUM,DATE,TIME,STREET1,STREET2,OFFSET,ROAD_CLASS,DISTRICT,...,REDLIGHT,ALCOHOL,DISABILITY,HOOD_158,NEIGHBOURHOOD_158,HOOD_140,NEIGHBOURHOOD_140,DIVISION,x,y
0,1,3389067,893184.0,1/1/2006 10:00:00 AM,236,WOODBINE AVE,O CONNOR DR,NaN,Major Arterial,Toronto and East York,...,NaN,Yes,NaN,60,Woodbine-Lumsden,60,Woodbine-Lumsden (60),D55,635468.3685,4.839881e+06
1,2,3389068,893184.0,1/1/2006 10:00:00 AM,236,WOODBINE AVE,O CONNOR DR,NaN,Major Arterial,Toronto and East York,...,NaN,Yes,NaN,60,Woodbine-Lumsden,60,Woodbine-Lumsden (60),D55,635468.3685,4.839881e+06
2,3,3389069,893184.0,1/1/2006 10:00:00 AM,236,WOODBINE AVE,O CONNOR DR,NaN,Major Arterial,Toronto and East York,...,NaN,Yes,NaN,60,Woodbine-Lumsden,60,Woodbine-Lumsden (60),D55,635468.3685,4.839881e+06
3,4,3389070,893184.0,1/1/2006 10:00:00 AM,236,WOODBINE AVE,O CONNOR DR,NaN,Major Arterial,Toronto and East York,...,NaN,Yes,NaN,60,Woodbine-Lumsden,60,Woodbine-Lumsden (60),D55,635468.3685,4.839881e+06
4,5,3389071,893184.0,1/1/2006 10:00:00 AM,236,WOODBINE AVE,O CONNOR DR,NaN,Major Arterial,Toronto and East York,...,NaN,Yes,NaN,60,Woodbine-Lumsden,60,Woodbine-Lumsden (60),D55,635468.3685,4.839881e+06


In [11]:
ksi_df.shape

(18957, 54)

In [12]:
ksi_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18957 entries, 0 to 18956
Data columns (total 54 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   OBJECTID           18957 non-null  int64  
 1   INDEX              18957 non-null  int64  
 2   ACCNUM             14027 non-null  float64
 3   DATE               18957 non-null  object 
 4   TIME               18957 non-null  int64  
 5   STREET1            18957 non-null  object 
 6   STREET2            17251 non-null  object 
 7   OFFSET             3820 non-null   object 
 8   ROAD_CLASS         18471 non-null  object 
 9   DISTRICT           18728 non-null  object 
 10  LATITUDE           18957 non-null  float64
 11  LONGITUDE          18957 non-null  float64
 12  ACCLOC             13501 non-null  object 
 13  TRAFFCTL           18882 non-null  object 
 14  VISIBILITY         18933 non-null  object 
 15  LIGHT              18953 non-null  object 
 16  RDSFCOND           189

In [13]:
summary_ksi_df = pd.DataFrame(
    {
        "Data Type": ksi_df.dtypes,
        "Non-Null Count": ksi_df.notnull().sum(),
        "Missing Count": ksi_df.isnull().sum(),
        "Missing %": (ksi_df.isnull().mean() * 100).round(2),
        "Unique Values": ksi_df.nunique(),
    }
)
summary_ksi_df

,Data Type,Non-Null Count,Missing Count,Missing %,Unique Values
OBJECTID,int64,18957,0,0.00,18957
INDEX,int64,18957,0,0.00,18957
ACCNUM,float64,14027,4930,26.01,4955
DATE,object,18957,0,0.00,4128
TIME,int64,18957,0,0.00,1341
STREET1,object,18957,0,0.00,1942
STREET2,object,17251,1706,9.00,2822
OFFSET,object,3820,15137,79.85,494
ROAD_CLASS,object,18471,486,2.56,11
DISTRICT,object,18728,229,1.21,4


In [16]:
# ranges and values of elements
print("Numeric Statistical Assessment")
ksi_df.describe().T

Numeric Statistical Assessment


,count,mean,std,min,25%,50%,75%,max
OBJECTID,18957.0,9.479000e+03,5.472559e+03,1.000000e+00,4.740000e+03,9.479000e+03,1.421800e+04,1.895700e+04
INDEX,18957.0,3.994181e+07,3.749731e+07,3.363207e+06,5.408154e+06,7.820656e+06,8.092292e+07,8.180752e+07
ACCNUM,14027.0,5.576714e+08,1.180026e+09,2.530100e+04,1.034561e+06,1.220029e+06,1.386958e+06,4.008024e+09
TIME,18957.0,1.364956e+03,6.313104e+02,0.000000e+00,9.240000e+02,1.450000e+03,1.852000e+03,2.359000e+03
LATITUDE,18957.0,4.371030e+01,5.629382e-02,4.358968e+01,4.366173e+01,4.370235e+01,4.375594e+01,4.385545e+01
LONGITUDE,18957.0,-7.939649e+01,1.044838e-01,-7.963839e+01,-7.947147e+01,-7.939739e+01,-7.931829e+01,-7.912297e+01
FATAL_NO,870.0,2.875287e+01,1.765700e+01,1.000000e+00,1.400000e+01,2.750000e+01,4.200000e+01,7.800000e+01
x,18957.0,6.291816e+05,8.364338e+03,6.096257e+05,6.231770e+05,6.291991e+05,6.354240e+05,6.510241e+05
y,18957.0,4.840951e+06,6.324585e+03,4.827328e+06,4.835532e+06,4.840070e+06,4.846036e+06,4.857444e+06


In [17]:

print("Categorical Statistical Assessment")
ksi_df.describe(include="object").T

categorical Statistical Assessment


,count,unique,top,freq
DATE,18957,4128,8/17/2014 8:00:00 AM,35
STREET1,18957,1942,YONGE ST,403
STREET2,17251,2822,BATHURST ST,156
OFFSET,3820,494,10 m West of,61
ROAD_CLASS,18471,11,Major Arterial,13376
DISTRICT,18728,4,Toronto and East York,6328
ACCLOC,13501,10,At Intersection,8774
TRAFFCTL,18882,10,No Control,9021
VISIBILITY,18933,8,Clear,16373
LIGHT,18953,9,Daylight,10779


# Handling Missing Values:
The missing-value stage is actually the trickiest part of this dataset, because most of the "missingness" here isn't random — it means something specific. Treating it carelessly (e.g. one global dropna() or filling everything with mode) will wreck other columns

**Group 1 — Binary involvement flags (huge missing %, but missing = "No")**
- PEDESTRIAN, CYCLIST, AUTOMOBILE, MOTORCYCLE, TRUCK, TRSN_CITY_VEH, EMERG_VEH, PASSENGER, SPEEDING, AG_DRIV, REDLIGHT, ALCOHOL, DISABILITY
- These columns only contain the value "Yes" when true, and NaN otherwise — there's no "No" recorded. 95%+ missing on ALCOHOL/DISABILITY doesn't mean "bad data," it means alcohol/disability wasn't a factor in 95% of collisions.
- Fix: fillna("No"), not drop, not mode-impute.

In [18]:
group1_cols = [
    "PEDESTRIAN",
    "CYCLIST",
    "AUTOMOBILE",
    "MOTORCYCLE",
    "TRUCK",
    "TRSN_CITY_VEH",
    "EMERG_VEH",
    "PASSENGER",
    "SPEEDING",
    "AG_DRIV",
    "REDLIGHT",
    "ALCOHOL",
    "DISABILITY",
]
ksi_df[group1_cols] = ksi_df[group1_cols].fillna("No")
print(ksi_df[group1_cols].isnull().sum())

PEDESTRIAN       0
CYCLIST          0
AUTOMOBILE       0
MOTORCYCLE       0
TRUCK            0
TRSN_CITY_VEH    0
EMERG_VEH        0
PASSENGER        0
SPEEDING         0
AG_DRIV          0
REDLIGHT         0
ALCOHOL          0
DISABILITY       0
dtype: int64


In [20]:
print(ksi_df["ALCOHOL"].value_counts())
print("\n", ksi_df["AG_DRIV"].value_counts())

ALCOHOL
No     18149
Yes      808
Name: count, dtype: int64

 AG_DRIV
Yes    9836
No     9121
Name: count, dtype: int64


**Group 2 — Role-conditional attributes (missing = not applicable to that record)**
- PEDTYPE, PEDACT, PEDCOND only apply when the involved person is a pedestrian; CYCLISTYPE, CYCACT, CYCCOND only apply to cyclists.
- Their ~83–96% missingness is because most records aren't pedestrians/cyclists at all.
- Fix: fillna("Not Applicable") — dropping rows would delete almost every non-pedestrian/non-cyclist record.

In [21]:
group2_cols = [
    # Pedestrian-specific features
    "PEDTYPE",
    "PEDACT",
    "PEDCOND",
    # Cyclist-specific features
    "CYCLISTYPE",
    "CYCACT",
    "CYCCOND",
]
ksi_df[group2_cols] = ksi_df[group2_cols].fillna("Not Applicable")
print(ksi_df[group2_cols].isnull().sum())

PEDTYPE       0
PEDACT        0
PEDCOND       0
CYCLISTYPE    0
CYCACT        0
CYCCOND       0
dtype: int64


In [27]:
print(ksi_df["PEDACT"].value_counts())
print("\n", ksi_df["CYCACT"].value_counts())

PEDACT
Not Applicable                           15730
Crossing with right of way                1019
Crossing, no Traffic Control               730
Crossing without right of way              453
On Sidewalk or Shoulder                    284
Other                                      240
Running onto Roadway                       235
Crossing, Pedestrian Crossover              68
Person Getting on/off Vehicle               39
Coming From Behind Parked Vehicle           37
Walking on Roadway with Traffic             33
Playing or Working on Highway               26
Walking on Roadway Against Traffic          24
Crossing marked crosswalk without ROW       21
Pushing/Working on Vehicle                  16
Person Getting on/off School Bus             2
Name: count, dtype: int64

 CYCACT
Not Applicable                  18155
Driving Properly                  444
Disobeyed Traffic Control          85
Other                              79
Failed to Yield Right of Way       65
Lost control    

Group 3 — Event-conditional
FATAL_NO (95.4% missing) **bold text**
- Only has a value for fatal injuries — it's a sequence number for fatalities in a multi-fatality collision.
- Missing = not fatal. fillna(0) or leave as "N/A" flag, not statistical imputation.
- Fix: Fill NaN Values with 0

In [28]:
ksi_df["FATAL_NO"] = ksi_df["FATAL_NO"].fillna(0).astype(int)
print("Missing count in FATAL_NO:", ksi_df["FATAL_NO"].isnull().sum())

Missing count in FATAL_NO: 0


In [29]:
print(ksi_df["FATAL_NO"].value_counts().head(10))

FATAL_NO
0     18087
11       20
1        16
3        16
4        16
5        16
2        16
6        16
7        16
12       16
Name: count, dtype: int64


**Group 4 — Genuinely missing/incomplete records (true missing data)**
- ACCLOC (28.8%), INITDIR (27.8%), ROAD_CLASS (2.6%), DISTRICT (1.2%), TRAFFCTL (0.4%), VISIBILITY, RDSFCOND, LIGHT, IMPACTYPE, INVTYPE, ACCLASS (all <1%)
— these are cases where the report just wasn't filled in completely. This is only genuine "missing data" group.
- Fix: The strategy is split into two parts:
  - A - dropping negligible missing rows ($< 3\%$)
  - B - imputing medium missingness (~$28\%$) as "Unknown"

In [30]:
group3A_cols = [
    "ACCLASS",
    "LIGHT",
    "VISIBILITY",
    "IMPACTYPE",
    "RDSFCOND",
    "TRAFFCTL",
    "DISTRICT",
    "ROAD_CLASS",
]
group3B_cols = ["ACCLOC", "INITDIR"]

ksi_df = ksi_df.dropna(subset=group3A_cols)
ksi_df[group3B_cols]= ksi_df[group3B_cols].fillna("Unkown")

In [33]:
print("Negligible Missing Rows\n",ksi_df[group3A_cols].isnull().sum())
print("\nMedium Missing Rows\n",ksi_df[group3B_cols].isnull().sum())

Negligible Missing Rows
 ACCLASS       0
LIGHT         0
VISIBILITY    0
IMPACTYPE     0
RDSFCOND      0
TRAFFCTL      0
DISTRICT      0
ROAD_CLASS    0
dtype: int64

Medium Missing Rows
 ACCLOC     0
INITDIR    0
dtype: int64


Group 5 — Identifiers, not analysis variables
OFFSET (79.8% missing), STREET2 (9%), ACCNUM (26%), VEHTYPE (18.4%) **bold text**
- These are location/ID descriptors.
- ACCNUM missingness usually reflects records where an official number wasn't assigned

In [36]:
group5_cols = [
    "OFFSET",
    "STREET2",
    "VEHTYPE"]

In [39]:
ksi_df[group5_cols]= ksi_df[group5_cols].fillna("Unkown")
ksi_df["ACCNUM"] = ksi_df["ACCNUM"].fillna(-1) #temporary placeholder

In [44]:
print(ksi_df[group5_cols].isnull().sum())
print("ACCNUM    ", ksi_df["ACCNUM"].isnull().sum())

OFFSET     0
STREET2    0
VEHTYPE    0
dtype: int64
ACCNUM     0
